ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java

```diff
@@ -279,6 +292,72 @@ a database ( directory )
     return 0;
   }
 
+  /**
+   * Cleanup/drop tables from the given database which are bootstrapped by input dump dir.
+   * @throws HiveException Failed to drop the tables.
+   * @throws IOException File operations failure.
+   * @throws InvalidInputException Invalid input dump directory.
+   */
+  private void bootstrapRollbackTask() throws HiveException, IOException, InvalidInputException {
+    Path bootstrapDirectory = new PathBuilder(work.bootstrapDumpToRollback)
+            .addDescendant(ReplUtils.INC_BOOTSTRAP_ROOT_DIR_NAME).build();
+    FileSystem fs = bootstrapDirectory.getFileSystem(conf);
+
+    if (!fs.exists(bootstrapDirectory)) {
+      throw new InvalidInputException("Input bootstrap dump directory to rollback doesn't exist: "
+              + bootstrapDirectory);
+    }
+
+    FileStatus[] fileStatuses = fs.listStatus(bootstrapDirectory, EximUtil.getDirectoryFilter(fs));
+    if ((fileStatuses == null) || (fileStatuses.length == 0)) {
+      throw new InvalidInputException("Input bootstrap dump directory to rollback is empty: "
+              + bootstrapDirectory);
+    }
+
+    if (StringUtils.isNotBlank(work.dbNameToLoadIn) && (fileStatuses.length > 1)) {
+      throw new InvalidInputException("Multiple DB dirs in the dump: " + bootstrapDirectory
+                      + " is not allowed to load to single target DB: " + work.dbNameToLoadIn);
+    }
+
+    for (FileStatus dbDir : fileStatuses) {
```

Comment:
Given the above two conditions there's going to be exactly one entry in the fileStatuses array. Why do we need a for loop here? We could just get that one entry into dbDir and write rest of the code without a loop?

Reply:
If work.dbNameToLoadIn is empty or null, then there can be multiple DB directories. So, it will be array in that case. We cannot avoid the loop.

Reply:
Thanks for the clarification. I didn't understand dbNameToLoadIn's role in the condition. We should add a testcase for this as well to make sure that the multiple db case is handled correctly.

Reply:
Currently, we don't support it fully. It is for Warehouse level dump+load where we don't support multi-DB load yet. This code is there for future use when that will be supported. I'm just leaving it as it is for now.

Reply:
Ok. Thanks.